File: app/src/main/java/intellij/class13/deliverableexercise/list/java/DigitalHouseManager.java
package intellij.class13.deliverableexercise.list.java;

import android.os.Build;

import java.util.ArrayList;
import java.util.List;

public class DigitalHouseManager {
    private List<Student> studentList = new ArrayList<>();
    private List<Teacher> teacherList = new ArrayList<>();
    private List<Course> courseList = new ArrayList<>();
    private List<Enrollment> enrollmentList = new ArrayList<>();

    public void registerCourse(String name, Integer courseCode, Integer maxQuantityStudents) {
        courseList.add(new Course(name, courseCode, maxQuantityStudents));
        System.out.println("Curso " + name + " registrado na Digital House.");
    }

    public void removeCourse(Integer courseCode) {
        for (Course c : courseList) {
            if (c.getCourseCode().equals(courseCode)) {
                System.out.println("Curso " + c.getName() + " removido.");
                courseList.remove(c);
            }
        }
    }

    public void registerAssistantTeacher(String name, String lastName, Integer teacherCode, Integer quantityMonitoringHours) {
        teacherList.add(new AssistantTeacher(name, lastName, 0, teacherCode, quantityMonitoringHours));
        System.out.println("Professor Adjunto '" + name + " " + lastName + "' registrado na Digital House.");
    }

    public void registerTitularTeacher(String name, String lastName, Integer teacherCode, String specialty) {
        teacherList.add(new TitularTeacher(name, lastName, 0, teacherCode, specialty));
        System.out.println("Professor Titular '" + name + " " + lastName + "' registrado na Digital House.");
    }


    public void removeTeacher(Integer teacherCode) {
        //Eu pensei no FOR, mas a IDE me deu Essa função, Yuri ou João poderiam me explicar
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            teacherList.removeIf(teacher -> teacher.getTeacherCode().equals(teacherCode));
        }
//        for (Teacher teacher : teacherList) {
//            if (teacher.getTeacherCode().equals(teacherCode)) {
//                teacherList.remove(teacher);
//            }
//        }
    }

    public void enrollingStudent(String name, String lastName, Integer studentCode) {
        studentList.add(new Student(name, lastName, studentCode));
    }

    public void enrollingStudent(Integer studentCode, Integer courseCode) {
        Course course = getCourseInList(courseCode, courseList);
        Student student = getStudentInList(studentCode, studentList);

        if (course.addStudent(student)) {
            enrollmentList.add(new Enrollment(student, course));
            System.out.println("Matrícula realizada. Aluno " + student.toString() + " adicionado no Curso " + course.getName());
        } else {
            System.out.println("Não foi possível realizar a matrícula. Curso " + course.getName() + " lotado.");
        }
    }

    public void allocateTeachers(Integer courseCode, Integer titularTeacherCode, Integer assistantTeacherCode) {
        Teacher titularTeacher = getTeacherInList(titularTeacherCode, teacherList);
        Teacher assistantTeacher = getTeacherInList(assistantTeacherCode, teacherList);
        Course course = getCourseInList(courseCode, courseList);

        course.setTitularTeacher((TitularTeacher) titularTeacher);
        course.setAssistantTeacher((AssistantTeacher) assistantTeacher);
        System.out.println("Professor Titular " + titularTeacher.toString() + " e Professor Adjunto " + assistantTeacher.toString() + " alocado no Curso " + course.getName());

    }

    public List<Student> getStudentList() {
        return studentList;
    }

    public void setStudentList(List<Student> studentList) {
        this.studentList = studentList;
    }

    public List<Teacher> getTeacherList() {
        return teacherList;
    }

    public void setTeacherList(List<Teacher> teacherList) {
        this.teacherList = teacherList;
    }

    public List<Course> getCourseList() {
        return courseList;
    }

    public void setCourseList(List<Course> courseList) {
        this.courseList = courseList;
    }

    public List<Enrollment> getEnrollmentList() {
        return enrollmentList;
    }

    public void setEnrollmentList(List<Enrollment> enrollmentList) {
        this.enrollmentList = enrollmentList;
    }

    private Course getCourseInList(Integer courseCode, List<Course> courseList) {
        Course course = null;
        for (Course c : courseList) {
            if (c.getCourseCode().equals(courseCode)) {
                course = c;
                break;
            }
        }
        return course;
    }

    private Student getStudentInList(Integer studentCode, List<Student> studentList) {
        Student student = null;
        for (Student s : studentList) {
            if (s.getStudentCode().equals(studentCode)) {
                student = s;
                break;
            }
        }
        return student;
    }

    private Teacher getTeacherInList(Integer teacherCode, List<Teacher> teacherList) {
        Teacher teacher = null;
        for (Teacher t : teacherList) {
            if (t.getTeacherCode().equals(teacherCode)) {
                teacher = t;
                break;
            }
        }
        return teacher;
    }
}

File: src/class13/deliverableexercise/kotlin/DigitalHouseManager.kt
package class13.deliverableexercise.kotlin

class DigitalHouseManager {
    private var studentList = mutableListOf<Student>()
    private var teacherList = mutableListOf<Teacher>()
    private var courseList = mutableListOf<Course>()
    private var enrollmentList = mutableListOf<Enrollment>()

    fun registerCourse(name: String, courseCode: Int, maxQuantityStudents: Int) {
        courseList.add(Course(name, courseCode, maxQuantityStudents))
        println("Curso $name registrado na Digital House.")
    }

    fun removeCourse(courseCode: Int) {
        for (course in courseList) {
            if (course.courseCode == courseCode) {
                println("Curso " + course.name + " removido.")
                courseList.remove(course)
            }
        }
    }

    fun registerAssistantTeacher(name: String, lastName: String, teacherCode: Int, quantityMonitoringHours: Int) {
        teacherList.add(AssistantTeacher(name, lastName, 0, teacherCode, quantityMonitoringHours))
        println("Professor Adjunto '$name $lastName' registrado na Digital House.")
    }

    fun registerTitularTeacher(name: String, lastName: String, teacherCode: Int, specialty: String) {
        teacherList.add(TitularTeacher(name, lastName, 0, teacherCode, specialty))
        println("Professor Titular '$name $lastName' registrado na Digital House.")
    }

    fun removeTeacher(teacherCode: Int?) {
        for (teacher in teacherList) {
            if (teacher.teacherCode == teacherCode) {
                teacherList.remove(teacher)
            }
        }
    }

    fun enrollingStudent(name: String, lastName: String, studentCode: Int) {
        studentList.add(Student(name, lastName, studentCode))
    }

    fun enrollingStudent(studentCode: Int, courseCode: Int) {
        val course = getCourseInList(courseCode, courseList)
        val student = getStudentInList(studentCode, studentList)
        if (course.addStudent(student)) {
            enrollmentList.add(Enrollment(student, course))
            println("Matrícula realizada. Aluno " + student.toString() + " adicionado no Curso " + course.name)
        } else {
            println("Não foi possível realizar a matrícula. Curso " + course.name + " lotado.")
        }
    }

    fun allocateTeachers(courseCode: Int, titularTeacherCode: Int, assistantTeacherCode: Int) {
        var titularTeacher = getTeacherInList(titularTeacherCode, teacherList)
        var assistantTeacher = getTeacherInList(assistantTeacherCode, teacherList)
        val course = getCourseInList(courseCode, courseList)
        course.apply {
            titularTeacher = (titularTeacher as TitularTeacher)
            assistantTeacher = (assistantTeacher as AssistantTeacher)
        }
        println("Professor Titular " + titularTeacher.toString() + " e Professor Adjunto " + assistantTeacher.toString() + " alocado no Curso " + course.name)
    }

    private fun getCourseInList(courseCode: Int, courseList: List<Course>): Course {
        lateinit var course: Course
        for (c in courseList) {
            if (c.courseCode == courseCode) {
                course = c
                break
            }
        }
        return course
    }

    private fun getStudentInList(studentCode: Int, studentList: List<Student>): Student {
        lateinit var student: Student
        for (s in studentList) {
            if (s.studentCode == studentCode) {
                student = s
                break
            }
        }
        return student
    }

    private fun getTeacherInList(teacherCode: Int, teacherList: List<Teacher>): Teacher {
        lateinit var teacher: Teacher
        for (t in teacherList) {
            if (t.teacherCode.equals(teacherCode)) {
                teacher = t
                break
            }
        }
        return teacher
    }
}